src/perl/lib/wtsi_clarity/epp/isc/beckman_file_generator.pm

```diff
@@ -0,0 +1,209 @@
+  isa => 'wtsi_clarity::util::beckman',
+  required => 0,
+  lazy_build => 1,
+);
+sub _build__beckman {
+  my $self = shift;
+  return wtsi_clarity::util::beckman->new();
+}
+
+has '_beckman_file' => (
+  is => 'ro',
+  isa => 'wtsi_clarity::util::textfile',
+  required => 0,
+  lazy_build => 1,
+);
+sub _build__beckman_file {
+  my $self = shift;
+  return $self->_beckman->get_file($self->internal_csv_output);
+}
+
+has 'internal_csv_output' => (
+  is => 'ro',
+  isa => 'ArrayRef',
+  required => 0,
+  lazy_build => 1,
+);
+sub _build_internal_csv_output {
+  croak '_build_internal_csv_output must be overriden'
+}
+
+override 'run' => sub {
+  my $self= shift;
+  super();
+
+  $self->_beckman_file->saveas(q{./} . $self->beckman_file_name);
+
+  return;
+};
+
+sub row {
+  my ($self, $dest_well, $analyte_data, $sample_num) = @_;
+  return c->new(@{$self->_beckman->headers})
+    ->reduce( sub {
+      my $method = $self->get_method_from_header($b);
+      my $value = $self->$method($dest_well, $analyte_data, $sample_num);
+      $a->{$b} = $value;
+      $a;
+    }, {});
+}
+
+## no critic(Subroutines::ProhibitUnusedPrivateSubroutines)
+sub _get_sample {
+  my ($self, $dest_well, $analyte_data, $sample_nr) = @_;
+  return $sample_nr;
+}
+
+sub _get_name {
+  my ($self, $dest_well, $analyte_data, $sample_nr) = @_;
+  return $DEFAULT_PLATE_NAME_PREFIX . q{1};
+}
+
+sub _get_source_ean13 {
+  my ($self, $dest_well, $analyte_data, $sample_nr) = @_;
+  return $DEFAULT_SOURCE_EAN13;
+}
+
+sub _get_source_barcode {
+  my ($self, $dest_well, $analyte_data, $sample_nr) = @_;
+  return $DEFAULT_SOURCE_BARCODE;
+}
+
+sub _get_source_stock {
+  my ($self, $dest_well, $analyte_data, $sample_nr) = @_;
+  return $DEFAULT_SOURCE_STOCK;
+}
+
+sub _get_destination_barcode {
+  my ($self, $dest_well, $analyte_data, $sample_nr) = @_;
+  return $DEFAULT_DEST_BARCODE;
+}
+
+sub _get_source_well {
+  my ($self) = @_;
+  return $self->_get_not_implemented_yet;
+}
+
+sub _get_destination_ean13 {
+  my ($self, $dest_well, $analyte_data, $sample_nr) = @_;
+  return $DEFAULT_DEST_EAN13;
+}
+
+sub _get_destination_well {
+  my ($self) = @_;
+  return $self->_get_not_implemented_yet;
+}
+
+sub _get_source_volume {
+  my ($self) = @_;
+  return $self->_get_not_implemented_yet;
+}
+
+sub _get_not_implemented_yet {
+  my ($self, $sample_id) = @_;
+  return qq{*} ; #qq{Not implemented yet};
+}
+
+1;
+
+__END__
+
+=head1 NAME
+
+wtsi_clarity::epp::isc::beckman_file_generator
+
+=head1 SYNOPSIS
+
+  package wtsi_clarity::epp::isc::some_process_that_needs_to_generate_a_beckman_robot_file
+  extends 'wtsi_clarity::epp::isc::beckman_file_generator'
+
+  overrides '_build_internal_csv_output' => sub {
+    ['here', 'is', 'some', 'sample', 'data']
+  };
+
+=head1 DESCRIPTION
+
+  Should be subclassed by any process that wants to create a beckman file. That subclass must implement
+  a _build_internal_csv_output method.
+
+  This class probably shouldn't be instantiated itself.
```

Comment:
I think this sentence is not sounds right to me: "This class probably shouldn't be instantiated itself." Then it should be an abstract class, I suppose, but it is not. From the maintenance perspective composition is better, because if you use inheritance, then you could have a problem, if you would like to change your super class and your subclass was using that feature you just changed.